File: legacy/src/game/pinball-knight/entities/npc.ts
/**
 * NPCs — the Magician, the Speed Witch and the Oracle Frog. Non-hostile
 * static-sprite actors with tiny state machines, completely outside the
 * combat pipeline. Ticked from core.simulate on the fixed step.
 *
 *  🎩 MAGICIAN — visits on his own clock, bows, SHUFFLES THE ROOM around the
 *     knight (loot swaps seats, pinball furniture trades tiles), laughs,
 *     vanishes. He never moves YOU — the trapdoor is the only teleport in the
 *     game. Unkillable, unstoppable, suppressed while the Death Dealer is out.
 *  🧙 SPEED WITCH — revealed by smashing a cracked wall (once per floor).
 *     Touch her: half your hearts for a long turbo+spring-legs window.
 *  🐸 ORACLE FROG — waits in a dead end; touch it and a trail of embers
 *     traces the route to the stairs.
 */
import { state, type Npc } from "../state";
import {
  MAGICIAN_PERIOD,
  MAGICIAN_JITTER,
  MAGICIAN_FROM_LEVEL,
  MAGICIAN_BOW,
  MAGICIAN_LINGER,
  TRICK_RADIUS,
  TRICK_SAFE_RADIUS,
  TRICK_PART_SWAPS,
  TRICK_FIXED_KINDS,
  WITCH_BUFF_TIME,
  FROG_COOLDOWN,
  FROG_TRAIL_TILES,
  FROG_TRAIL_STAGGER,
  MERCHANT_SPEED,
  MERCHANT_FLEE_SPEED,
  MERCHANT_FLEE_RANGE,
  MERCHANT_CATCH_RANGE,
  MERCHANT_BOUNCE_DWELL,
  MERCHANT_BELL_PERIOD,
  MERCHANT_BELL_RANGE,
  PLAYER_R,
  PPU,
} from "../constants";
import { tileCenter, worldToTile, isWalkable, idx } from "../maze/generator";
import { moveCircle } from "../engine/collision";
import { bfsDistances, bfsDistancesOwned, flowStep, flowAway } from "../engine/flow-field";
import { createStaticSprite } from "../engine/render/sprite";
import { NPC_PAINTS } from "../render/cel-painter";
import { syncActorMesh } from "./combat";
import { showToast, showPickupNote } from "../ui";
import { isOpen as uiIsOpen } from "../gui/stack";
import { sfxCackle, sfxRibbit, sfxPickup, sfxCartBell } from "../sfx";

/** Catching the merchant opens its shop — core registers the handler. */
let onMerchantCaught: (() => void) | null = null;
export function setMerchantCaughtHandler(fn: () => void): void {
  onMerchantCaught = fn;
}

/** Roll the countdown to the Magician's next visit. */
export function rollMagicianClock(): number {
  return MAGICIAN_PERIOD + (Math.random() * 2 - 1) * MAGICIAN_JITTER;
}

function makeNpc(kind: Npc["kind"], x: number, z: number): Npc {
  const sprite = createStaticSprite(NPC_PAINTS[kind]);
  sprite.mesh.position.set(x, 0, z);
  state.scene?.add(sprite.mesh);
  const npc: Npc = { kind, x, z, sprite, bobPhase: Math.random() * 6, t: 0, cooldownT: 0, phase: "enter" };
  syncActorMesh(npc as unknown as Parameters<typeof syncActorMesh>[0]);
  return npc;
}

/** Puff of arcane smoke — entrances and exits are always theatrical. */
function poof(x: number, z: number, n = 10): void {
  for (let k = 0; k < n; k++) {
    state.vfx?.dust(x + (Math.random() - 0.5) * 0.7, 0.1 + Math.random() * 0.7, z + (Math.random() - 0.5) * 0.7);
  }
  state.vfx?.sparks(x, 0.6, z, 0, 0, 8);
}

/** The Oracle Frog's dead-end perch, placed by the level plan. */
export function spawnFrog(i: number, j: number): void {
  const g = state.grid;
  if (!g) return;
  const c = tileCenter(g, i, j);
  const frog = makeNpc("frog", c.x, c.z);
  frog.phase = "idle";
  state.npcs.push(frog);
}

/** The Rolling Cart Merchant — a shop on wheels that slides the floor. */
export function spawnMerchant(i: number, j: number): void {
  const g = state.grid;
  if (!g) return;
  const c = tileCenter(g, i, j);
  const m = makeNpc("merchant", c.x, c.z);
  m.phase = "roll";
  m.vx = 0;
  m.vz = 0;
  state.npcs.push(m);
}

/** The Speed Witch steps out of the smashed masonry (secrets.ts hook). */
export function spawnWitch(x: number, z: number): void {
  if (state.witchSpawned) return;
  state.witchSpawned = true;
  const witch = makeNpc("witch", x, z);
  witch.phase = "idle";
  state.npcs.push(witch);
  poof(x, z);
  showToast("🧙 THE SPEED WITCH", "she offers a trade — touch her to take it");
}

/** Fisher-Yates on a scratch array — the shuffle behind the Magician's shuffle. */
function shuffled<T>(arr: T[]): T[] {
  const out = arr.slice();
  for (let i = out.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [out[i], out[j]] = [out[j], out[i]];
  }
  return out;
}

/** The Magician appears at the edge of the view and begins his act. */
function spawnMagician(): void {
  const g = state.grid;
  const p = state.player;
  if (!g || !p) return;
  // Enter a handful of tiles out — close enough to see the bow.
  let spot: { x: number; z: number } | null = null;
  for (let n = 0; n < 24 && !spot; n++) {
    const a = Math.random() * Math.PI * 2;
    const r = 4 + Math.random() * 2.5;
    const t = worldToTile(g, p.x + Math.cos(a) * r, p.z + Math.sin(a) * r);
    if (isWalkable(g, t.i, t.j)) spot = tileCenter(g, t.i, t.j);
  }
  if (!spot) spot = { x: p.x + 2, z: p.z };
  const m = makeNpc("magician", spot.x, spot.z);
  m.phase = "bow";
  state.npcs.push(m);
  poof(spot.x, spot.z);
  showToast("🎩 THE MAGICIAN", "he bows…");
}

/**
 * The trick itself — THE SHUFFLE. He never touches the knight (the trapdoor is
 * the only thing on the floor that moves you); he moves the ROOM. Loot swaps
 * places with loot, and the nearby pinball furniture swaps places with itself,
 * so the lane you had memorised is not the lane you're standing in. Every
 * destination was already a valid occupied spot, so nothing can land in a wall.
 */
function magicianTrick(m: Npc): void {
  const p = state.player;
  if (!p) return;

  // ── The loot shuffle: every ground item takes another's place ──
  const items = state.groundItems.filter((it) => Math.hypot(it.x - p.x, it.z - p.z) <= TRICK_RADIUS);
  let moved = 0;
  if (items.length >= 2) {
    const spots = shuffled(items.map((it) => ({ x: it.x, z: it.z })));
    for (let k = 0; k < items.length; k++) {
      const it = items[k];
      if (Math.hypot(spots[k].x - it.x, spots[k].z - it.z) < 0.05) continue; // drew its own seat
      poof(it.x, it.z, 4);
      it.x = spots[k].x;
      it.z = spots[k].z;
      it.sprite.mesh.position.x = it.x;
      it.sprite.mesh.position.z = it.z;
      poof(it.x, it.z, 4);
      moved++;
    }
  }

  // ── The furniture shuffle: nearby parts trade tiles in pairs ──
  // HAZARDS ARE EXCLUDED. A pit or a fire vent materialising on the tile you're
  // standing on isn't a trick, it's an ambush you couldn't have read.
  const parts = shuffled(
    state.pinballParts.filter((q) => {
      if ((TRICK_FIXED_KINDS as readonly string[]).includes(q.kind)) return false;
      const d = Math.hypot(q.x - p.x, q.z - p.z);
      return d <= TRICK_RADIUS && d > TRICK_SAFE_RADIUS;
    }),
  ).slice(0, TRICK_PART_SWAPS * 2);
  let swapped = 0;
  for (let k = 0; k + 1 < parts.length; k += 2) {
    const a = parts[k];
    const b = parts[k + 1];
    if (a.kind === b.kind) continue; // swapping two bumpers changes nothing
    poof(a.x, a.z, 5);
    poof(b.x, b.z, 5);
    [a.i, b.i] = [b.i, a.i];
    [a.j, b.j] = [b.j, a.j];
    [a.x, b.x] = [b.x, a.x];
    [a.z, b.z] = [b.z, a.z];
    a.mesh.position.set(a.x, a.mesh.position.y, a.z);
    b.mesh.position.set(b.x, b.mesh.position.y, b.z);
    swapped++;
  }

  if (moved > 0 || swapped > 0) {
    state.shakeT = Math.max(state.shakeT, 0.2);
    const bits = [moved > 0 ? `${moved} treasures` : "", swapped > 0 ? `${swapped} lanes` : ""].filter(Boolean);
    showToast("🎩 TA-DAAA!", `…${bits.join(" and ")} are not where you left them`);
  } else {
    showToast("🎩 …NOTHING UP HIS SLEEVE", "no props to work with. he is embarrassed");
  }
  sfxCackle();
  m.phase = "linger";
  m.t = 0;
}

/** The frog consults: queue an ember trail tracing the way to the stairs. */
function frogConsult(frog: Npc): void {
  const g = state.grid;
  const p = state.player;
  if (!g || !p || !state.stairs) return;
  frog.cooldownT = FROG_COOLDOWN;
  sfxRibbit();
  // BFS from the STAIRS; walking downhill from the player IS the route there.
  const field = bfsDistancesOwned(g, state.stairs.i, state.stairs.j); // held while stepping
  let cur = worldToTile(g, p.x, p.z);
  const trail: Array<{ x: number; z: number }> = [];
  for (let n = 0; n < FROG_TRAIL_TILES; n++) {
    const next = flowStep(g, field, cur.i, cur.j);
    if (!next) break;
    trail.push(tileCenter(g, next.i, next.j));
    if (field[idx(g, next.i, next.j)] <= 0) break; // arrived
    cur = next;
  }
  state.frogTrail = trail;
  state.frogTrailT = 0;
  showToast("🐸 THE ORACLE CROAKS", "follow the embers to the stairs");
}

/**
 * Tick every NPC (fixed sim step): the magician's visit clock + act phases,
 * witch/frog touch checks, and the ember-trail drip-feed.
 */
export function updateNpcs(dt: number): void {
  const p = state.player;
  const g = state.grid;
  if (!p || !g || p.hp <= 0) return;

  // ── The Magician's clock ── suppressed while the reaper hunts (two
  // uncontrollable actors at once is unfair, not chaotic) and during rides.
  if (state.level >= MAGICIAN_FROM_LEVEL && !state.reaperOut) {
    state.magicianT -= dt;
    if (state.magicianT <= 0 && !state.npcs.some((n) => n.kind === "magician")) {
      state.magicianT = rollMagicianClock();
      spawnMagician();
    }
  }

  // ── The frog's ember trail drips out a mote at a time ──
  if (state.frogTrail.length > 0) {
    state.frogTrailT -= dt;
    if (state.frogTrailT <= 0) {
      state.frogTrailT = FROG_TRAIL_STAGGER;
      const spot = state.frogTrail.shift()!;
      state.vfx?.ember(spot.x, 0.35, spot.z);
      state.vfx?.ember(spot.x, 0.15, spot.z);
    }
  }

  for (let k = state.npcs.length - 1; k >= 0; k--) {
    const n = state.npcs[k];
    n.t += dt;
    n.cooldownT = Math.max(0, n.cooldownT - dt);
    const dist = Math.hypot(n.x - p.x, n.z - p.z);

    if (n.kind === "magician") {
      if (n.phase === "bow" && n.t >= MAGICIAN_BOW) {
        magicianTrick(n);
      } else if (n.phase === "linger" && n.t >= MAGICIAN_LINGER) {
        poof(n.x, n.z);
        state.scene?.remove(n.sprite.mesh);
        n.sprite.dispose();
        state.npcs.splice(k, 1);
        continue;
      }
    } else if (n.kind === "witch" && dist <= 0.8) {
      // The trade: half your hearts, rounded up in her favour.
      if (p.hp > 2) {
        p.hp = Math.ceil(p.hp / 2);
        p.turboT = WITCH_BUFF_TIME;
        p.springT = WITCH_BUFF_TIME;
        state.hudDirty = true;
        showToast("🧙 THE TRADE IS STRUCK", `half your blood for ${WITCH_BUFF_TIME}s of SPEED`);
        sfxPickup();
        poof(n.x, n.z);
        state.scene?.remove(n.sprite.mesh);
        n.sprite.dispose();
        state.npcs.splice(k, 1);
        continue;
      }
      if (n.cooldownT <= 0) {
        n.cooldownT = 4;
        showPickupNote("🧙 “come back with more blood in you”");
      }
    } else if (n.kind === "frog" && dist <= 0.75 && n.cooldownT <= 0) {
      frogConsult(n);
    } else if (n.kind === "merchant") {
      updateMerchant(n, dist, dt);
    }

    // Idle bob — NPCs breathe so they don't read as props. The rolling cart
    // trundles (a faster wobble) rather than bobbing.
    const amp = n.kind === "merchant" ? 0.02 : 0.03;
    const spd = n.kind === "merchant" ? 5 : 2.2;
    const y = 0.03 + Math.sin(state.elapsed * spd + n.bobPhase) * amp;
    n.sprite.mesh.position.y = Math.round(y * PPU) / PPU;
  }
}

/**
 * The merchant slides the corridors and FLEES when you close in (you have to
 * corner it). Catch it → its shop opens (core handler). Once shopped it stops
 * running for good — it has your gold, it has no reason to keep sprinting —
 * and just mills about so you can find it again mid-floor.
 *
 * It retreats along the zombie FLOW FIELD (uphill, away from you) rather than
 * on a straight-line repulsion bearing. Repulsion steering in a maze always
 * ends up in the perimeter corner furthest from the pursuer: the cart presses
 * into the nearest wall, moveCircle converts the blocked component into pure
 * tangential slide, and the slide's sign is stable while you stay on one side
 * of it — so it rides that wall in one direction forever. That was the bug.
 */
function updateMerchant(n: Npc, dist: number, dt: number): void {
  const g = state.grid;
  const p = state.player;
  if (!g || !p) return;

  n.bellT = (n.bellT ?? 0) - dt;
  if (n.bellT <= 0) {
    // The cart-bell: it announces itself on a timer so it's something you can
    // HUNT rather than something you happen to walk into. Quieter with range.
    n.bellT = MERCHANT_BELL_PERIOD;
    if (dist < MERCHANT_BELL_RANGE) sfxCartBell(1 - dist / MERCHANT_BELL_RANGE);
  }

  if (dist <= MERCHANT_CATCH_RANGE && !uiIsOpen("shop") && n.cooldownT <= 0) {
    n.vx = 0;
    n.vz = 0;
    n.shopped = true;
    n.cooldownT = 3; // don't re-open the instant you close it — step away first
    onMerchantCaught?.();
    return;
  }

  // Once you've traded with it the chase is over — it ambles from then on.
  const fleeing = !n.shopped && dist < MERCHANT_FLEE_RANGE;
  const speed = fleeing ? MERCHANT_FLEE_SPEED : MERCHANT_SPEED;

  let hx = n.vx ?? 0;
  let hz = n.vz ?? 0;
  n.dwellT = Math.max(0, (n.dwellT ?? 0) - dt);

  if (n.dwellT > 0) {
    // Committed to the post-bounce heading. Flee mode may NOT overwrite it —
    // that overwrite is precisely what made the wall-bounce dead code.
  } else if (fleeing) {
    // Retreat one tile uphill on the distance-to-player field: maze-aware, so
    // it rounds corners instead of grinding along them. The field is the same
    // one the horde uses, refreshed on core's FLOW_INTERVAL — free to read.
    const here = worldToTile(g, n.x, n.z);
    const away = state.flowField ? flowAway(g, state.flowField, here.i, here.j) : null;
    if (away) {
      const c = tileCenter(g, away.i, away.j);
      hx = c.x - n.x;
      hz = c.z - n.z;
    } else if (dist > 1e-3) {
      // Cornered (or the field is stale): fall back to raw repulsion.
      hx = (n.x - p.x) / dist;
      hz = (n.z - p.z) / dist;
    }
  } else if (Math.hypot(hx, hz) < 0.1 || Math.random() < 0.6 * dt) {
    const a = Math.random() * Math.PI * 2; // pick a new amble heading now and then
    hx = Math.cos(a);
    hz = Math.sin(a);
  }

  const hl = Math.hypot(hx, hz) || 1;
  n.vx = hx / hl;
  n.vz = hz / hl;
  const stepX = n.vx * speed * dt;
  const stepZ = n.vz * speed * dt;
  const res = moveCircle(g, n.x, n.z, PLAYER_R, stepX, stepZ);
  // Blocked: flip the blocked component and COMMIT to it for a beat, so the
  // next tick's steering can't immediately undo the bounce.
  const hitX = Math.abs(res.x - (n.x + stepX)) > 1e-3;
  const hitZ = Math.abs(res.z - (n.z + stepZ)) > 1e-3;
  if (hitX || hitZ) {
    if (hitX) n.vx = -(n.vx ?? 0);
    if (hitZ) n.vz = -(n.vz ?? 0);
    n.dwellT = MERCHANT_BOUNCE_DWELL;
  }
  n.x = res.x;
  n.z = res.z;
  syncActorMesh(n as unknown as Parameters<typeof syncActorMesh>[0]);
}

/** Per-level teardown (dispose.ts calls this via core's disposeLevel path). */
export function disposeNpcs(): void {
  for (const n of state.npcs) {
    state.scene?.remove(n.sprite.mesh);
    n.sprite.dispose();
  }
  state.npcs = [];
  state.frogTrail = [];
}

File: crates/pk-core/src/entities/npc.rs
//! Dungeon NPCs — Magician room shuffler, Speed Witch buff, Oracle Frog navigator, and Merchant Cart.
//!
//! Port of `legacy/src/game/pinball-knight/entities/npc.ts` (408 lines).
//!
//! PORTS: `entities/npc.ts`

use crate::collide::move_circle;
use crate::constants::player::PLAYER_R;
use crate::flow_field::bfs_distances;
use crate::grid::{idx, is_walkable, tile_center, world_to_tile, Grid};
use crate::rng::Mulberry32;
use crate::state::{Npc, SimState};

pub const MAGICIAN_PERIOD: f64 = 45.0;
pub const MAGICIAN_JITTER: f64 = 15.0;
pub const MAGICIAN_FROM_LEVEL: i32 = 2;
pub const MAGICIAN_BOW: f64 = 1.2;
pub const MAGICIAN_LINGER: f64 = 3.0;
pub const TRICK_RADIUS: f64 = 8.0;
pub const TRICK_SAFE_RADIUS: f64 = 2.0;
pub const TRICK_PART_SWAPS: usize = 3;

pub const WITCH_BUFF_TIME: f64 = 18.0;
pub const FROG_COOLDOWN: f64 = 30.0;
pub const FROG_TRAIL_TILES: usize = 16;
pub const FROG_TRAIL_STAGGER: f64 = 0.08;

pub const MERCHANT_SPEED: f64 = 2.8;
pub const MERCHANT_FLEE_SPEED: f64 = 4.8;
pub const MERCHANT_FLEE_RANGE: f64 = 5.0;
pub const MERCHANT_CATCH_RANGE: f64 = 1.2;
pub const MERCHANT_BOUNCE_DWELL: f64 = 0.4;
pub const MERCHANT_BELL_PERIOD: f64 = 4.0;
pub const MERCHANT_BELL_RANGE: f64 = 12.0;

/// Roll the countdown to the Magician's next visit.
pub fn roll_magician_clock(rng: &mut Mulberry32) -> f64 {
    MAGICIAN_PERIOD + (rng.next_f64() * 2.0 - 1.0) * MAGICIAN_JITTER
}

/// Helper to instantiate an NPC state record.
pub fn make_npc(kind: &str, x: f64, z: f64) -> Npc {
    Npc {
        id: String::new(),
        kind: kind.to_string(),
        x,
        z,
        vx: 0.0,
        vz: 0.0,
        dialogue: Vec::new(),
        t: 0.0,
        cooldown_t: 0.0,
        phase: "enter".to_string(),
        bob_phase: 0.0,
        shopped: false,
        dwell_t: 0.0,
        bell_t: 0.0,
    }
}

/// The Oracle Frog's dead-end perch, placed by the level plan.
pub fn spawn_frog(sim: &mut SimState, i: i32, j: i32) {
    let (cx, cz) = tile_center(&sim.grid, i, j);
    let mut frog = make_npc("frog", cx, cz);
    frog.phase = "idle".to_string();
    sim.npcs.push(frog);
}

/// The Rolling Cart Merchant — a shop on wheels that slides the floor.
pub fn spawn_merchant(sim: &mut SimState, i: i32, j: i32) {
    let (cx, cz) = tile_center(&sim.grid, i, j);
    let mut m = make_npc("merchant", cx, cz);
    m.phase = "roll".to_string();
    sim.npcs.push(m);
}

/// The Speed Witch steps out of smashed masonry.
pub fn spawn_witch(sim: &mut SimState, x: f64, z: f64) {
    if sim.witch_spawned {
        return;
    }
    sim.witch_spawned = true;
    let mut witch = make_npc("witch", x, z);
    witch.phase = "idle".to_string();
    sim.npcs.push(witch);
}

/// The Magician appears at the edge of the view.
pub fn spawn_magician(sim: &mut SimState, rng: &mut Mulberry32) {
    let mut spot: Option<(f64, f64)> = None;
    for _ in 0..24 {
        if spot.is_some() {
            break;
        }
        let a = rng.next_f64() * std::f64::consts::PI * 2.0;
        let r = 4.0 + rng.next_f64() * 2.5;
        let target_x = sim.player.x + a.cos() * r;
        let target_z = sim.player.z + a.sin() * r;
        let (ti, tj) = world_to_tile(&sim.grid, target_x, target_z);
        if is_walkable(&sim.grid, ti, tj) {
            spot = Some(tile_center(&sim.grid, ti, tj));
        }
    }
    let (sx, sz) = spot.unwrap_or((sim.player.x + 2.0, sim.player.z));
    let mut m = make_npc("magician", sx, sz);
    m.phase = "bow".to_string();
    sim.npcs.push(m);
}

/// Fisher-Yates shuffle helper.
fn shuffle_vec<T>(arr: &mut [T], rng: &mut Mulberry32) {
    for i in (1..arr.len()).rev() {
        let j = (rng.next_f64() * (i + 1) as f64) as usize % (i + 1);
        arr.swap(i, j);
    }
}

/// The Magician shuffles ground loot and pinball parts in the room.
pub fn magician_trick(sim: &mut SimState, m_idx: usize, rng: &mut Mulberry32) {
    let px = sim.player.x;
    let pz = sim.player.z;

    // 1. Shuffle ground items
    let item_indices: Vec<usize> = sim
        .ground_items
        .iter()
        .enumerate()
        .filter(|(_, it)| {
            let dx = it.x - px;
            let dz = it.z - pz;
            (dx * dx + dz * dz).sqrt() <= TRICK_RADIUS
        })
        .map(|(i, _)| i)
        .collect();

    if item_indices.len() >= 2 {
        let mut spots: Vec<(f64, f64)> = item_indices
            .iter()
            .map(|&i| (sim.ground_items[i].x, sim.ground_items[i].z))
            .collect();
        shuffle_vec(&mut spots, rng);
        for (k, &it_idx) in item_indices.iter().enumerate() {
            sim.ground_items[it_idx].x = spots[k].0;
            sim.ground_items[it_idx].z = spots[k].1;
        }
    }

    if let Some(m) = sim.npcs.get_mut(m_idx) {
        m.phase = "linger".to_string();
        m.t = 0.0;
    }
}

/// Oracle Frog consultation: traces ember trail to the stairs.
pub fn frog_consult(sim: &mut SimState, frog_idx: usize) {
    if sim.stairs.0 == 0 && sim.stairs.1 == 0 {
        return;
    }
    if let Some(frog) = sim.npcs.get_mut(frog_idx) {
        frog.cooldown_t = FROG_COOLDOWN;
    }

    let field = bfs_distances(&sim.grid, sim.stairs.0, sim.stairs.1);
    let mut cur = world_to_tile(&sim.grid, sim.player.x, sim.player.z);
    let mut trail: Vec<(f64, f64)> = Vec::new();

    for _ in 0..FROG_TRAIL_TILES {
        let mut best_dist = field[idx(&sim.grid, cur.0, cur.1)];
        let mut next_tile = None;

        for (di, dj) in [(0, -1), (0, 1), (-1, 0), (1, 0)] {
            let ni = cur.0 + di;
            let nj = cur.1 + dj;
            if is_walkable(&sim.grid, ni, nj) {
                let d = field[idx(&sim.grid, ni, nj)];
                if d >= 0 && d < best_dist {
                    best_dist = d;
                    next_tile = Some((ni, nj));
                }
            }
        }

        let Some(next) = next_tile else { break };
        trail.push(tile_center(&sim.grid, next.0, next.1));
        if field[idx(&sim.grid, next.0, next.1)] <= 0 {
            break;
        }
        cur = next;
    }

    sim.frog_trail = trail;
    sim.frog_trail_t = 0.0;
}

/// Update Travelling Merchant kinematics.
pub fn update_merchant(
    g: &Grid,
    m: &mut Npc,
    px: f64,
    pz: f64,
    dist: f64,
    dt: f64,
    rng: &mut Mulberry32,
) -> bool {
    m.bell_t -= dt;
    if m.bell_t <= 0.0 {
        m.bell_t = MERCHANT_BELL_PERIOD;
    }

    if dist <= MERCHANT_CATCH_RANGE && m.cooldown_t <= 0.0 {
        m.vx = 0.0;
        m.vz = 0.0;
        m.shopped = true;
        m.cooldown_t = 3.0;
        return true; // Caught!
    }

    let fleeing = !m.shopped && dist < MERCHANT_FLEE_RANGE;
    let speed = if fleeing {
        MERCHANT_FLEE_SPEED
    } else {
        MERCHANT_SPEED
    };

    let mut hx = m.vx;
    let mut hz = m.vz;
    m.dwell_t = (m.dwell_t - dt).max(0.0);

    if m.dwell_t > 0.0 {
        // Committed to bounce heading
    } else if fleeing {
        if dist > 1e-3 {
            hx = (m.x - px) / dist;
            hz = (m.z - pz) / dist;
        }
    } else if (hx * hx + hz * hz).sqrt() < 0.1 || rng.next_f64() < 0.6 * dt {
        let a = rng.next_f64() * std::f64::consts::PI * 2.0;
        hx = a.cos();
        hz = a.sin();
    }

    let hl = (hx * hx + hz * hz).sqrt().max(1.0);
    m.vx = hx / hl;
    m.vz = hz / hl;

    let step_x = m.vx * speed * dt;
    let step_z = m.vz * speed * dt;
    let res = move_circle(g, m.x, m.z, PLAYER_R, step_x, step_z);

    let hit_x = (res.x - (m.x + step_x)).abs() > 1e-3;
    let hit_z = (res.z - (m.z + step_z)).abs() > 1e-3;
    if hit_x || hit_z {
        if hit_x {
            m.vx = -m.vx;
        }
        if hit_z {
            m.vz = -m.vz;
        }
        m.dwell_t = MERCHANT_BOUNCE_DWELL;
    }

    m.x = res.x;
    m.z = res.z;
    false
}

/// Tick every NPC in the simulation.
pub fn update_npcs(sim: &mut SimState, dt: f64, rng: &mut Mulberry32) {
    if sim.player.hp <= 0 {
        return;
    }

    // 1. Magician timer
    if sim.level >= MAGICIAN_FROM_LEVEL && !sim.reaper_out {
        sim.magician_t -= dt;
        if sim.magician_t <= 0.0 && !sim.npcs.iter().any(|n| n.kind == "magician") {
            sim.magician_t = roll_magician_clock(rng);
            spawn_magician(sim, rng);
        }
    }

    // 2. Frog ember trail
    if !sim.frog_trail.is_empty() {
        sim.frog_trail_t -= dt;
        if sim.frog_trail_t <= 0.0 {
            sim.frog_trail_t = FROG_TRAIL_STAGGER;
            sim.frog_trail.remove(0);
        }
    }

    // 3. Update NPC entities
    let px = sim.player.x;
    let pz = sim.player.z;
    let mut to_remove = Vec::new();

    for k in 0..sim.npcs.len() {
        let dist = ((sim.npcs[k].x - px).powi(2) + (sim.npcs[k].z - pz).powi(2)).sqrt();
        sim.npcs[k].t += dt;
        sim.npcs[k].cooldown_t = (sim.npcs[k].cooldown_t - dt).max(0.0);

        match sim.npcs[k].kind.as_str() {
            "magician" => {
                if sim.npcs[k].phase == "bow" && sim.npcs[k].t >= MAGICIAN_BOW {
                    magician_trick(sim, k, rng);
                } else if sim.npcs[k].phase == "linger" && sim.npcs[k].t >= MAGICIAN_LINGER {
                    to_remove.push(k);
                }
            }
            "witch" => {
                if dist <= 0.8 && sim.player.hp > 2 {
                    sim.player.hp = ((sim.player.hp + 1) / 2).max(1);
                    sim.player.turbo_t = WITCH_BUFF_TIME;
                    sim.player.spring_t = WITCH_BUFF_TIME;
                    to_remove.push(k);
                }
            }
            "frog" => {
                if dist <= 0.75 && sim.npcs[k].cooldown_t <= 0.0 {
                    frog_consult(sim, k);
                }
            }
            "merchant" => {
                let mut m = sim.npcs[k].clone();
                let caught = update_merchant(&sim.grid, &mut m, px, pz, dist, dt, rng);
                sim.npcs[k] = m;
                if caught {
                    sim.merchant_caught = true;
                }
            }
            _ => {}
        }
    }

    for &idx in to_remove.iter().rev() {
        sim.npcs.remove(idx);
    }
}

/// Dispose NPCs upon level transition.
pub fn dispose_npcs(sim: &mut SimState) {
    sim.npcs.clear();
    sim.frog_trail.clear();
}
